directories.md

```diff
@@ -0,0 +1,160 @@
+
+Using one-hot encoded values is a very common pattern in systems programming because it allows for efficient usage of resources while preserving the readability of the code. For example, let's suppose we wanted to write a function to create a file with some permissions, one option would be for the prototype of the function to be as follows:
+
+	int create_file(char *filename, int user-read, int user-write, int user-execute, ...);
+  
+which would result in the function being called as `create_file("dummyfile", 1, 1, 0, ...)` 
+
+If instead we relied on the one-hot encoded macros, our function could simply have the prototype:
+
+	int create_file(char *filename, int mode);
+  
+and could be called as `create_file("dummyfile", S_IRUSR | S_IWUSR)` and it would carry the same information while using considerably less space and mantain sufficient readability.
+
+## lstat
+
+int lstat(const char *restrict pathname,
+                struct stat *restrict statbuf);
+
+`lstat` performs an identical function to `stat()` with the only diference being that if `pathname` refers to a symbolic link, the information on `statbuf` will represent information about the link itself, and not the file pointed to by the link.
+
+In the following scenario:
+
+	$ ls -l
+	drwxrwx---  2 xurxo xurxo        3 Mar 25 17:38 dir/
+	lrwxrwxrwx  1 xurxo xurxo        3 Mar 25 17:39 dirlink -> dir/
+  
+`stat("dir", statbuf)` would be equivalent to `stat("dirlink", statbuf)` but not equivalent to `lstat("dirlink", statbuf)`
+
+
+### Exercises
+
+(1.1) Using the permission macros, write an expression using bitwise OR to represent the following permission string `rw-rw-r--`
+
+(1.2) Using the permission macros, write an expression to verify that a given mode does not have write permissions for any group.
+
+(1.3) T/F We can modify the parameters to readdir to prevent it from returning the directory entries for "." and ".."
+
+(1.4) Does the following code contain any memory leaks? If so, how would you fix it?
+
+	int main(int argc, char **argv){
+	    DIR *dir = opendir(argv[1]);
+	    if (dir != NULL)
+	        printf("Opened directory\n");
+	     else
+	        printf("Could not open directory\n");
+	}
+
+(1.5) How many times will `readdir()` execute in the following directory before it returns `NULL`. You may assume it never fails
+
+	$ ls -l
+	-rw-rw----  1 xurxo xurxo   0 Mar 25 17:58 file1.txt
+	-rw-rw----  1 xurxo xurxo   0 Mar 25 17:58 file2.txt
+	lrwxrwxrwx  1 xurxo xurxo   4 Mar 25 17:38 link -> link
+	drwxrwx---  2 xurxo xurxo   2 Mar 25 17:58 subdir/
+
+Check the additional, `permstat.c` exercise on CourseWorks for more hands-on coding practice. 
+
+### Solutions
+
+(1.1) `(S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH)`
+
+(1.2) `(mode & (S_IWUSR | S_IWGRP | S_IWOTH)) == 0`
+
+(1.3) False.
+
+(1.4) Yes, it can be prevented calling `closedir()`
+
+(1.5) 6
```

Comment:
```suggestio
### Solutions
<details><summary>Click Here for Solutions</summary>
(1.1) `(S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH)`

(1.2) `(mode & (S_IWUSR | S_IWGRP | S_IWOTH)) == 0`

(1.3) False.

(1.4) Yes, it can be prevented calling `closedir()`

(1.5) 6
</details>
```

Reply:
might have to actually check that this works but ideally solutions should be hidden